### PR TITLE
Custom Functor Typeclass

### PR DESCRIPTION
**Pull request by fosskers**
### TODO
- [x] `Functor` trait
- [x] Instance for `TileLayerMetadata[K]` (maps over `KeyBounds`)
- [x] Remove `SpatialComponent` constraint on `TIleLayerMetadata.apply`
- [x] Get `.map` to inject properly (thanks to @pomadchin )
- [x] ~~Add test to confirm #1575 fix~~ (no point, the compiler guarantees it)

### Motivation
#1575 describes how `TileLayerRDD.toSpatial` doesn't reflect the change in key type in its metadata. After much deliberation, we recognized our attempts to address this as attempts to write a `Functor` instance for `TileLayerMetadata[K]` which maps over its `KeyBounds`. Unfortunately, Scalaz's `Functor` trait is a bit too restrictive, in that types that use it can't impose additional constraints on their inner type. Compare Scalaz:

```scala
trait Functor[F[_]] extends InvariantFunctor[F] { self =>

  /** Lift `f` into `F` and apply to `F[A]`. */
  def map[A, B](fa: F[A])(f: A => B): F[B]
...
}
```
to our implementation:

```scala
 trait Functor[F[_], A] extends MethodExtensions[F[A]]{
   /** Lift `f` into `F` and apply to `F[A]`. */
   def map[B](f: A => B): F[B]
 }
```
where `A` is lifted into `Functor`'s type signature. This allows us to write code like this:
```scala
def foo[K: A: B: C](implicit f: Functor[TileLayerMetadata, K]): ??? = {
...
  tlm.map(k => ... ) // `k` has all the power of its constraints A, B, and C.
...
}
```

### CODE REVIEW

**Comment by fosskers** [comment]
This may or may not be an actual Functor, depending on how well we can convince ourselves that the second Functor law holds:

```haskell
fmap (g . h) = (fmap g) . (fmap h)
```

If you only had one implicit `Functor[F, A]` in scope, then:

```scala
val m: F[A] = ???
val f: A => B = ???
val g: B => C = ???

val n: F[C] = m.map(f).map(g)
```
wouldn't compile. But if you *did* have both `Functor[F, A]` and `Functor[F, B]` (which may both use the same underlying implicit class, see `class TlmFunctor[A]`), then the above would compile. Furthermore, so would:

```scala
val n: F[C] = m.map(k => g(f(k)))
```
and so the Functor law holds, and it is indeed a Functor.

File: spark/src/main/scala/geotrellis/spark/filter/Implicits.scala [hunk 1]
```diff
@@ -16,9 +16,8 @@ trait Implicits {
       extends TileLayerRDDFilterMethods[K, V, M]
 
   implicit class withSpaceTimeToSpatialMethods[
-    K: SpatialComponent: TemporalComponent,
+    K: SpatialComponent: TemporalComponent: λ[α => M[α] => Functor[M, α]]: λ[α => Component[M[α], Bounds[α]]],
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/filter/Implicits.scala]: This should be type aliased w/ a broken down explained in a comment, pretty tough to parse and also understand why they just can't be context bounds on M

File: spark/src/main/scala/geotrellis/spark/package.scala [hunk 2]
```diff
@@ -62,6 +61,14 @@ package object spark
       new ContextRDD(rdd, metadata)
   }
 
+  implicit class TileLayerMetadataFunctor[A](val self: TileLayerMetadata[A]) extends Functor[TileLayerMetadata, A] {
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/package.scala]: Comments to explain why this is needed.

File: spark/src/main/scala/geotrellis/spark/filter/Implicits.scala (hunk 1 above)
Comment (fosskers) [inline spark/src/main/scala/geotrellis/spark/filter/Implicits.scala]: It's definitely space voodoo the first time you see it, yeah. For technical reasons they can't be bounds on M, so I'll explain that.

**Pull request merged**